main.cpp

```diff
@@ -6,134 +6,714 @@
+        return *this;
+    }
+    *value -= *rhs.value;
+    return *this;
+}
+
+DoubleType& DoubleType::subtract(const IntType& rhs)
+{
+    if (rhs.value == nullptr)
+    {
+        std::cout << "value null ptr\n";
+        return *this;
+    }
+    *value -= static_cast<double>(*rhs.value);
+    return *this;
+}
+
+DoubleType& DoubleType::multiply(const double& rhs)
+{
+    *value *= rhs;
+    return *this;
+}
+
+DoubleType& DoubleType::multiply(const FloatType& rhs)
+{
+    if (rhs.value == nullptr)
+    {
+        std::cout << "value null ptr\n";
+        return *this;
+    }
+    *value *= static_cast<double>(*rhs.value);
+    return *this;
+}
+
+DoubleType& DoubleType::multiply(const DoubleType& rhs)
+{
+    if (rhs.value == nullptr)
+    {
+        std::cout << "value null ptr\n";
+        return *this;
+    }
+    *value *= *rhs.value;
+    return *this;
+}
+
+DoubleType& DoubleType::multiply(const IntType& rhs)
+{
+    if (rhs.value == nullptr)
+    {
+        std::cout << "value null ptr\n";
+        return *this;
+    }
+    *value *= static_cast<double>(*rhs.value);
+    return *this;
+}
+
+DoubleType& DoubleType::divide(const double& rhs)
+{
+    if (rhs == 0.0)
+    {
+        std::cout << "divide by zero\n";
+        return *this;
+    }
+    *value /= rhs;
+    return *this;
+}
+
+DoubleType& DoubleType::divide(const FloatType& rhs)
 {
-    FloatType ft;
-    DoubleType dt;
-    IntType it;
+    if (rhs.value == nullptr)
+    {
+        std::cout << "value null ptr\n";
+        return *this;
+    }
+    if (*rhs.value == 0.f)
+    {
+        std::cout << "divide by zero\n";
+        return *this;
+    }
+    *value /= static_cast<double>(*rhs.value);
+    return *this;
+}
 
-    std::cout << "result of ft.add(): " << ft.add(1.0f, 3.0f) << '\n';
-    std::cout << "result of ft.subtract(): " << ft.subtract(1.0f, 3.0f) << '\n';
-    std::cout << "result of ft.multiply(): " << ft.multiply(2.0f, 3.0f) << '\n';
-    std::cout << "result of dt.add(): " << dt.add(1.0, 3.0) << "\n";
-    std::cout << "result of dt.divide(): " << dt.divide(1.0, 3.0) << "\n";
-    std::cout << "result of it.add(): " << it.add(100, 3) << "\n";
+DoubleType& DoubleType::divide(const DoubleType& rhs)
+{
+    if (rhs.value == nullptr)
+    {
+        std::cout << "value null ptr\n";
+        return *this;
+    }
+    if (*rhs.value == 0.0)
+    {
+        std::cout << "divide by zero\n";
+        return *this;
+    }
+    *value /= *rhs.value;
+    return *this;
+}
 
-    std::cout << "result of ft.divide(): " << ft.divide(1, 0) << '\n';
-    std::cout << "result of dt.divide(): " << dt.divide(1, 0) << '\n';
-    std::cout << "result of it.divide(): " << it.divide(1, 0) << '\n';
+DoubleType& DoubleType::divide(const IntType& rhs)
+{
+    if (rhs.value == nullptr)
+    {
+        std::cout << "value null ptr\n";
+        return *this;
+    }
+    if (*rhs.value == 0)
+    {
+        std::cout << "divide by zero\n";
+        return *this;
```

Comment:
- [ ] floating point division by 0 is not an error.  you don't need to return yet.